### ISSUE TITLE
چطوری میشه فیلتر نویسی کرد

### ISSUE THREAD

**Issue by engAmirEng** [mahs4d/tsetmc-api#20]
سلام
چطوری میشه فیلتری ک آماده دارم رو اعمال کنم؟

**Comment by alizadehmehdad691-bit** [comment]
from tsetmc_api.symbol import Symbol
import pandas as pd
import numpy as np

# --------------------------------------
# Indicator calculation functions
# --------------------------------------

def ema(series, period):
    return series.ewm(span=period, adjust=False).mean()

def rsi(series, period=14):
    delta = series.diff()
    gain = np.where(delta > 0, delta, 0)
    loss = np.where(delta < 0, -delta, 0)

    gain_roll = pd.Series(gain).rolling(window=period).mean()
    loss_roll = pd.Series(loss).rolling(window=period).mean()

    rs = gain_roll / loss_roll
    return 100 - (100 / (1 + rs))

# --------------------------------------
# Fetch symbol data from TSETMC
# --------------------------------------

symbol_id = "46348559193224090"   # example: خودرو
sym = Symbol(symbol_id)

history = sym.get_daily_history_data()

df = pd.DataFrame(history)
df = df.rename(columns={
    "final": "Close",
    "high": "High",
    "low": "Low",
    "open": "Open",
    "volume": "Volume",
    "value": "Value"
})

df["Date"] = pd.to_datetime(df["date"])
df = df.sort_values("Date")

# --------------------------------------
# Calculate indicators
# --------------------------------------

df["EMA20"] = ema(df["Close"], 20)
df["EMA50"] = ema(df["Close"], 50)
df["RSI14"] = rsi(df["Close"], 14)
df["VolMA20"] = df["Volume"].rolling(window=20).mean()

# --------------------------------------
# Buy signal conditions
# --------------------------------------

cond_price_above_ema20 = df["Close"] > df["EMA20"]
cond_trend_up = df["EMA20"] > df["EMA50"]

df["RSI_prev"] = df["RSI14"].shift(1)
cond_rsi_cross_50 = (df["RSI_prev"] <= 50) & (df["RSI14"] > 50)

cond_volume_high = df["Volume"] > df["VolMA20"]

df["BuySignal"] = (
    cond_price_above_ema20 &
    cond_trend_up &
    cond_rsi_cross_50 &
    cond_volume_high
)

# --------------------------------------
# Print output
# --------------------------------------

print(df[["Date", "Close", "EMA20", "EMA50", "RSI14", "Volume", "VolMA20", "BuySignal"]].tail(30))

buy_points = df[df["BuySignal"] == True]
print("\nBuy signals found:", len(buy_points))
print(buy_points[["Date", "Close", "EMA20", "EMA50", "RSI14", "Volume"]])
